admin/tools/provision/cloudmanager.py

```diff
@@ -0,0 +1,311 @@
+            time.sleep(5)
+            status = self.nova.images.get(qserv_image).status
+        logging.info("SUCCESS: Qserv image '{}' is active".format(self.snapshot_name))
+
+    def nova_servers_create(self, instance_id, userdata):
+        """
+        Boot an instance and check status
+        """
+        instance_name = "{0}-qserv-{1}".format(self._safe_username, instance_id)
+        logging.info("Launch an instance {}".format(instance_name))
+
+        # Launch an instance from an image
+        instance = self.nova.servers.create(name=instance_name,
+                                            image=self.image,
+                                            flavor=self.flavor,
+                                            userdata=userdata,
+                                            key_name=self.key,
+                                            nics=self.nics)
+        # Poll at 5 second intervals, until the status is no longer 'BUILD'
+        status = instance.status
+        while status == 'BUILD':
+            time.sleep(5)
+            instance.get()
+            status = instance.status
+        logging.info("status: {}".format(status))
+        logging.info("Instance {} is active".format(instance_name))
+
+        return instance
+
+    def detect_end_cloud_config(self, instance):
+        """
+        Add clean wait for cloud-init completion
+        """
+        check_word = "---SYSTEM READY FOR SNAPSHOT---"
+        end_word = None
+        while not end_word:
+            time.sleep(10)
+            output = instance.get_console_output()
+            logging.debug("console output: {}".format(output))
+            logging.debug("instance: {}".format(instance))
+            end_word = re.search(check_word, output)
+
+    def nova_servers_delete(self, server):
+        """
+        Shut down and delete a server
+        """
+        server.delete()
+
+    def manage_ssh_key(self):
+        """
+        Upload ssh public key
+        """
+        logging.info('Manage ssh keys: {}'.format(self.key))
+        if self.nova.keypairs.findall(name=self.key):
+            logging.debug('Remove previous ssh keys')
+            self.nova.keypairs.delete(key=self.key)
+
+        with open(os.path.expanduser(self.key_filename + ".pub")) as fpubkey:
+            self.nova.keypairs.create(name=self.key, public_key=fpubkey.read())
+
+    def get_floating_ip(self):
+        """
+        Allocate floating ip address to project
+        """
+        i = 0
+        floating_ips = self.nova.floating_ips.list()
+        floating_ip = None
+        floating_ip_pool = self.nova.floating_ip_pools.list()[0].name
+
+        # Check for available public ip address in project
+        while i < len(floating_ips) and floating_ip is None:
+            if floating_ips[i].instance_id is None:
+                floating_ip = floating_ips[i]
+                logging.debug('Available floating ip found {}'.format(floating_ip))
+            i += 1
```

Comment:
Using indices is a bit fragile and ugly, I think more idiomatic way to do it is like:

    for ip in self.nova.floating_ips.list():
        if ip.instance_id is None:
            floating_ip = ip
            logging.debug('Available floating ip found %s', floating_ip)
            break